client/containers/Pub/SpubHeader/SpubHeaderToolBar/SpubHeaderToolbar.tsx

```diff
@@ -0,0 +1,105 @@
+import React, { Dispatch, SetStateAction, useMemo } from 'react';
+import { Tab, Tabs, TabId, Icon, IconName, Button } from '@blueprintjs/core';
+import Color from 'color';
+
+import { GridWrapper } from 'components';
+import { SubmissionStatus } from 'types';
+import { usePageContext, usePendingChanges } from 'utils/hooks';
+
+require('./spubHeaderToolbar.scss');
+
+const renderTabTitle = (icon: IconName, title: string) => (
+	<>
+		<Icon icon={icon} /> {title}
+	</>
+);
+
+type Props = {
+	selectedTab: TabId;
+	onSelectTab: Dispatch<SetStateAction<TabId>>;
```

Comment:
So `Dispatch<SetStateAction<TabId>>` is the type of the `useState<TabId>` setter, but isn't what you want here, because it rejects callback functions that aren't `useState` setters. Let me break that down.

`SetStateAction<T>` is:
```ts
T | ((currentValue: T) => T)
```
Which is the set of things that the setter function from a `useState` hook accepts...e.g. you can say `setCount(3)` or `setCount(n => n + 1)`. Then, `Dispatch<T>` is just:
```ts
(value: T) => void
```
Together these correctly type the `useState` setter function. But in the component that calls this, we'll likely use a a callback like this to actually trigger a state change:

```ts
// In SpubHeader.tsx
const { updateSubmissionState } = usePubContext();

const setCurrentTab = useCallback((tabId: TabId) => {
    updateSubmissionState({ tabId });
}, [updateSubmissionState]);
```
The type of this function is `Dispatch<TabId>`, and TypeScript will not accept it as a prop here because it doesn't accept an argument like `(currentTab: TabId) => TabId`.

My head is spinning a bit writing all of this, but the long and short is that we don't want to assume that callback props are `useState` setters, because those functions have extra abilities (taking an updater function) that typical callbacks don't. Instead you could use one of these:

- `Dispatch<TabId>`
- `Callback<TabId>` (importing the `Callback<T>` utility from `types`)
- `(t: TabId) => unknown`

Any of these are correct, but the latter two are most common in our codebase.

Reply:
Callback functions have different abilities than useState functions. we want to use a callback because its like a Philips screwdriver, not a flathead.

Reply:
Good analogy. I think in this case `Dispatch<SetStateAction<TabId>>` is the phillips and `Callback<TabId>` is the flathead (React's `Dispatch<T>` and our `Callback<T>` are identical for our purposes). You can unwrap that to say that `SetStateAction<TabId>` is the phillips screwdriver and `TabId` is the flathead — you can use a `TabId` anywhere you can use a `SetStateAction<TabId>` but not vice-versa.

(it's nearly midnight so I reserve the right to wake up in the morning and realize that this is completely backwards 🤡)

Reply:
If we're talking _screws_, though, then `TabId` is the phillips. You can use it anywhere you have a flathead or phillips _screwdriver_ handy, whereas if you have a flathead screw you need a flathead screwdriver.

This is a pretty interesting analogy for the relationship between values and types.